### PR TITLE
Batched & chunked prefill

### PR DESCRIPTION
**Pull request by lucasavila00**
Continuing https://github.com/EricLBuehler/mistral.rs/pull/219

Closes https://github.com/EricLBuehler/mistral.rs/issues/216

### CODE REVIEW

File: mistralrs-core/src/xlora_models/mistral.rs [hunk 1]
```diff
@@ -525,26 +525,20 @@ impl XLoraModel {
...
+        let t = tgt_len;
+        let u = seqlen_offset + tgt_len;
+        let mask: Vec<_> = (0..t)
             .flat_map(|i| {
-                (0..tgt_len).map(move |j| {
-                    if i < j || j + sliding_window < i {
-                        f32::NEG_INFINITY
-                    } else {
+                (0..u).map(move |j| {
+                    if j + t + self.sliding_window.unwrap_or(tgt_len + 1) > i + u {
```
Comment (lucasavila00) [inline mistralrs-core/src/xlora_models/mistral.rs:533]: I'm not sure the sliding window part is right

File: mistralrs-core/src/xlora_models/mixtral.rs [hunk 2]
```diff
@@ -675,25 +675,20 @@ impl XLoraModel {
...
+        let t = tgt_len;
+        let u = seqlen_offset + tgt_len;
+        let mask: Vec<_> = (0..t)
             .flat_map(|i| {
-                (0..tgt_len).map(move |j| {
-                    if i < j || j + self.sliding_window < i {
-                        f32::NEG_INFINITY
-                    } else {
+                (0..u).map(move |j| {
+                    if j + t + self.sliding_window > i + u {
```
Comment (lucasavila00) [inline mistralrs-core/src/xlora_models/mixtral.rs:683]: I'm not sure the sliding window part is right

File: mistralrs-core/src/pipeline/mod.rs [hunk 3]
```diff
@@ -363,11 +366,45 @@ pub trait NormalModel {
...
+    pub positions: Vec<usize>,
+    pub positions_kernel: Tensor, // [bs, seq len]
+    pub context_lens: Vec<usize>,
+}
+
+fn calculate_inputs_prompt_batched(
+    seq: &mut Sequence,
+    device: &Device,
+    chunk_size: usize,
+) -> Result<Vec<InputMetadata>> {
```
Comment (lucasavila00) [inline mistralrs-core/src/pipeline/mod.rs:381]: Is it worth it making `InputMetadata` public? Or should this function return `ModelInputs`?

File: mistralrs-core/src/pipeline/mod.rs [hunk 4]
```diff
@@ -219,6 +219,9 @@ pub trait Pipeline: Send + Sync {
     fn get_max_seq_len(&self) -> usize;
     fn is_xlora(&self) -> bool;
     fn has_no_kv_cache(&self) -> bool;
+    fn get_prefill_chunk_size(&self) -> usize {
+        512
+    }
```
Comment (lucasavila00) [inline mistralrs-core/src/pipeline/mod.rs:224]: Should this be configurable?

This value should be as large as possible for performance, under the constraint that it doesn't OOM the system.

File: mistralrs-core/src/xlora_models/phi3.rs [hunk 5]
```diff
@@ -460,16 +460,12 @@ impl Model {
         tgt_len: usize,
         seqlen_offset: usize,
```
Comment (lucasavila00) [inline mistralrs-core/src/xlora_models/phi3.rs:461]: This function already passed the `seqlen_offset`. I think it was already being correctly calculated for the current use-case, but I'm not sure

File: mistralrs-core/src/xlora_models/gemma.rs [hunk 6]
```diff
@@ -561,16 +561,12 @@ impl XLoraModel {
         tgt_len: usize,
         seqlen_offset: usize,
```
Comment (lucasavila00) [inline mistralrs-core/src/xlora_models/gemma.rs:562]: This function already passed the `seqlen_offset`. I think it was already being correctly calculated for the current use-case, but I'm not sure

**Comment by lucasavila00** [comment]
@EricLBuehler I'd appreciate help on testing this.

It changed a lot of models, some of them I have never used.

I did test the quantized llama I usually run a lot though. Both on `mistral-bench` and talking to it with the interactive mode.

**Comment by EricLBuehler** [comment]
@lucasavila00, absolutely, I'll test the models out.

**Pull request closed without merging**